<h1>
    <br style ="color:#EA7600"> UCL </br> 
    </p>
</h1>

<div>

  <h4>
      <p style="font-size:18pt">COMP0163 Blockchain Technologies</p>
      <p style="font-size:18pt">Lab Practical 1</p>
      <p style="font-size:18pt">Part1</p>
  </h4>

</div>

# Before you start
* Install Anaconda on you PC. Anaconda is a powerful distribution of Python, which simplifies package management and deployment. Follow the instructions on the website:
> https://www.anaconda.com/download
* After installing, run the command below in the terminal to create the Anaconda environment for the coursework
> conda env create -f blockchain_env.yml
* After creating, run the command below in the terminal to activate the environment
> conda activate blockchain_env
* After activating, select the blockchain_env as your Kernel and you are good to go
> In Visual Studio Code, click on the "Select Kernel" tab located in the upper right corner. Then, click on "Python Environment", and finally choose "blockchain_env".





# Table of Contents
* [1. What is Blockchain?](#1.-What-is-Blockchain?)
* [2. Blockchain infraestructure](#2-Blockchain-infraestructure)
* [3. Blocks](#3-Blocks)
    * [3.1 The Block header](#3.1-The-Block-header)
    * [3.2 Linking blocks](#3.2-Linking-blocks)
    * [3.3 Signing data in blockchain](#3.3-Signing-data-in-blockchain)
    * [3.4 Merkle hash trees](#3.4-Merkle-hash-trees)
    * [3.5 From linked list to blockchain](#3.5-From-linked-list-to-blockchain)
* [4. Hashes](#4-Hashes)
    * [4.1 How do hashes work?](#4.1-How-do-hashes-work?)
    * [4.2 Proof-of-Work](#4.2-Proof-of-Work)
    * [4.3 Consensus](#4.3-Consensus)
* [5. Cryptography](#5.-Cryptography)

# 1. What is Blockchain?

Although blockchain has a variety of definitions, a blockchain can best be described as a data structure of blocks that are chained together to form a collection of records, called a ledger, with cryptography being a key ingredient in the process. A blockchain doesn't have a storage mechanism; instead, it has a set of protocols that govern the way in which information is forged. So, a blockchain can be stored in flat files or in a database.

Blockchain technology gained popularity due to the fact that its integrity can't easily be compromised. A compromised blockchain can be recognized for what it is, and rejected quite easily by anyone in a network. This integrity is achieved by cryptography, which is what binds the blocks together.

Blockchain's promise of providing such robust integrity is what eventually paved the way for the idea of sharing chains of data in untrusted peer-to-peer (P2P) networks. Validation of the blocks in a blockchain is what makes sure that a blockchain has a valid global state that can be accepted by everyone. Due to a blockchain's ability to share information in an open P2P network without any central authority governing it, the technology can have many different applications; however, the technology could not simply just be deployed to these applications immediately without any troubleshooting. Although blockchain technology, from the beginning, had a huge role to play in the decentralization of applications, it still faced several challenges with regards to its application in trustless environments. One of the biggest challenges was keeping a blockchain consistent across all the participants of a P2P network. This was solved by creating a consensus algorithm, which agrees on how the blocks should be appended to grow the chain in a trustless environment.

The term blockchain actually entails a number of concepts, including P2P network management, consensus mechanism, and more, all contributing to the creation of a decentralized application.

<img src="./figures/dlt_features.png" width="95%"  align="center">

Before we move on, let's have a look at several definitions of the word, blockchain:
> *A blockchain is a peer-to-peer distributed ledger forged by consensus, combined with a system for "smart contracts" and other assistive technologies.* **hyperledger.org**

> *A blockchain is a specific form or subset of distributed ledger technologies, which constructs a chronological chain of blocks, hence the name "block-chain."*- **Antony Lewis**, the Director of Research at R3

> *The blockchain data structure is an ordered, back-linked list of blocks.* - **Andreas Antonopoulos**, a popular Bitcoin evangelist

# 2. Blockchain infraestructure

<img src="./figures/database_vs_decentralized.png" width="70%"  align="center">

A Blockchain network essentially manages a `decentralized database`. The concept was first described in 2009 [in a cryptography blog](https://bitcoin.org/bitcoin.pdf) by a person or group of persons with the pseudonym `“Satoshi Nakamoto.”` Nakamoto proposed an  innovative peer-to-peer (P2P) electronic currency called Bitcoin that would enable online payments to be transferred directly from sender to recipient without an intermediary [(NAKAMOTO 2008)](https://bitcoin.org/bitcoin.pdf). 

Although the Bitcoin currency is an interesting and innovative experiment (by January 2019 worth more than US Dollar (USD) 130 bn in value - see https://coinmarketcap.com/ as of 01/2019), it was the trust-building Blockchain technology (i.e., cryptographic approach) behind it that soon began to be seen as the actual revolutionary aspect. 

Blockchain is to Bitcoin what the Internet is to email. It is a big electronic system based on distributed ledgers, on top of which you can build applications. Currency (Bitcoin) is just one of these applications.

<img src="./figures/blockchain_infraestructure.png" width="70%"  align="center">

The infrastructure of a Blockchain network consists of distributed computer servers (called nodes, validators, miners, etc.). These servers operate under the rule that interactions are only made by servers permanently following rules of the protocols. Intermediaries (e.g., banks, clearing houses, trading platforms, centralized service providers, etc.) are hence no longer needed.

Types of blockchains:

- `Private blockchain (Permissioned)`: only authorized entities can hold a copy of the ledger or participate in transactions. Here, the nodes know each other, at least to a certain degree. The consensus process is either controlled by a group of known participants ( within conglomerates) or by a single participant (e.g., a company or a governmental agency), which adds some degree of centralization. E.g. [Hyperledger](https://www.hyperledger.org/), [iota](https://www.iota.org/), and [ripple](https://ripple.com/)

    Private blockchains are suitable for organizations where a ledger only needs to be shared internally. Permissioned blockchains are often mutable or not strictly immutable, and their transactions can be modified with some effort; this is in stark contrast to public blockchains, where this is nearly impossible. Permissioned blockchains are still decentralized ledgers, but they will have some nodes with limited capability within the organization, whereas nodes in public blockchains are treated impartially.


- `Public blockchain (Permission-less)`: Permission-less Blockchain networks (e.g., Bitcoin and Ethereum) are open to everyone. Anyone  may become a ledger node (i.e., hold a copy of the ledger) and add valid transaction entries.


- `Consortium blockchain`: a hybrid blockchain that is semi-decentralized. It combines the best features of both permissionless and permissioned blockchains. Instead of assigning most tasks to a single organization, a consortium blockchain assigns the same tasks to nodes maintained by multiple organizations. Instead of having a single validator node, there can be multiple nodes. Although a consortium blockchain is permissioned, it's more decentralized than a private blockchain. E.g. [Facebook's Libra](https://libra.org/en-US/)


# 3. Blocks

If we consider a blockchain as a data structure, then the `blocks` are aggregated sets of data that are used for formation of the blockchain. *Blockchain formation is similar to linked list formation, where each node has a reference to the next node in the sequence*. In the case of blockchain, each `block` has a reference to the previous node, thus forming a link all the way to the initial block (known as the `genesis block`) of the chain.

Bitcoin was what really gave birth to blockchain technology, and since then, several new applications of the distributed blockchain database have emerged, most notably during early 2014. This era is widely known as the era of blockchain 2.0. `Namecoin` was one of the first concepts to emerge. It expanded the scope of Bitcoin's blockchain and was introduced as a distributed naming system based on Bitcoin. However, unlike Bitcoin, it was able to store data as **key-value pairs** in the public blockchain. This concept influenced a number of blockchain 2.0 applications that would go on to gain popularity.  

A blockchain can be stored in either flat file or database format. Bitcoin uses [LevelDB](https://github.com/google/leveldb) to store metadata about all the blocks that are downloaded to the disk.

Permissioned and permissionless blockchains, for instance, have slight variations in their block structure. We are going tp be using Bitcoin's permissionless block structure as a reference to try and identify its characteristics:

<img src="./figures/blocks_structure.png" height="30%" align="center">

**Block Header** and **Transactions** are the most important parts of the block, as they are responsible for the hash value, which is the identity of the block. The **Block Size** is the size of the entire block. The **Block Header** contains all the metadata of the block, and the **Transaction Counter** has the count of transactions. Finally, all the **Transactions** are stored in the block.

A blockchain starts with an initial block called the `genesis block`. If the chain is traversed backward from any given block, it will end up at the `genesis block`, proving that the entire chain is legitimate and valid. The `genesis block` is often statically coded in a public or permissionless blockchain, but it's created by the first participant in the case of the permissioned blockchain.

## 3.1 The Block header

The block header, as stated before, consists of the metadata of the block. This holds the information that's needed to link the blocks in the blockchain:

<img src="./figures/block_header.png" height="30%" align="center">

Each block header will have the components outlined in the image above. These are the minimum fields required in a permissionless blockchain, such as Bitcoin, to efficiently create a block that can be appended to an immutable blockchain. The **Previous Block Hash** field is a reference to the last block created. The **Merkle Root** is the value of the Merkle hash tree; it summarizes all the transactions in the block. **Timestamp**, **Difficulty Target**, and **Nonce** are used by the PoW consensus algorithm to solve the hash puzzle.

## 3.2 Linking blocks

Blocks are linked in a blockchain using references, just like in a linked list, but here the blocks are linked by referencing the hash value (identifier) of the previous block. Each full node in a blockchain network will maintain a complete blockchain and append a new block whenever it has one to append. Due to the decentralized nature of the blockchain, each node will verify the block before linking it to the local blockchain record.

The computed hash value of each block is the combination of the hash of the previous block and its own block data. This results in a dependency between neighboring blocks and nearly unbreakable links:

<img src="./figures/blocks_linking.png" height="30%" align="center">

[Satoshi]((https://bitcoin.org/bitcoin.pdf)) explained how the concept of timestamping should be used. All the items are hashed and the block is timestamped, meaning that the subsequent block will include this timestamp, creating an ordered chain of blocks.

Each node in the blockchain network follows a simple process for appending new blocks to its existing local blockchain. Whenever a node receives a block from the network, it checks for the previous block hash. If the hash value matches with the hash value of the last block on the node's local blockchain, then the node accepts this block and appends it to the current blockchain. As long as this is the longest known blockchain, the blocks would be considered valid by all the peers in the network in a PoW-based blockchain.

As previously explained, aside from the list of transactions, *we also put the parent's hash in the block*. But for now, let's just use a simple ID to indicate the parent instead of using a hash. Parent id is just the previous block id. Here, think of the stack. In the beginning, there is no block. Instead, we put **Block A**, which has three transactions: 
- Transaction 1
- Transaction 2
- Transaction 3

Since **Block A** is the first block, it has no parent. **Block A** is the equivalent to the `genesis block` in Bitcoin.

We then apply **Block B** to **Block A**, which consists of two transactions: **Transaction 4** and **Transaction 5**. **Block B** is not the first one in this blockchain. Consequently, we set the parent section in **Block B** as the **Block A** id because **Block A** is the parent of **Block B**. Then, we put **Block C** in the blockchain, which has two transactions: **Transaction 6** and **Transaction 7**.

The parent section in **Block C** would be the **Block B** id, and so on. To simplify things, we increment the id from 0 by 1 for every new block:

<img src="./figures/blocks.png" width="50%"  align="center">

Let's implement a database to record the history of what people like and hate. This means that when you said you like cats at one point in history, you won't be able to change that history. You may add new history when you change your mind (for example, if you then hate cats), but that won't change the fact that you liked them in the past. So, we can see that in the past you liked cats, but now you hate them. We want to make this database full of integrity and secure against cheating. Take a look at the following code block:

In [1]:
import hashlib
import json

# Let's define a class to link the blocks as lists
class Block:
    id = None
    history = None
    parent_id = None
    parent_hash = None

# Let's create block A and store the values in it
block_A = Block()

# This would the equivalent of normally putting the hash, but for simplicity and for better understanding
# we are using simple ids
block_A.id = 1

# This is equivalent to a transaction stored in the block
block_A.history = 'Nelson likes cat'

block_B = Block()
block_B.id = 2
block_B.history = 'Marie likes dog'

# Appending the id of the block A (genesis block) to the id of block B
block_B.parent_id = block_A.id

# Serializing the data of Block A into json format and hashing it to add it as parent hash to block B
block_B.parent_hash = hashlib.sha256(json.dumps(block_A.__dict__).encode('utf-8')).hexdigest()

block_C = Block()
block_C.id = 3
block_C.history = 'Sky hates dog'
block_C.parent_id = block_B.id

If you studied computer science, you will recognize this data structure, which is called a linked list. Now, there is a problem. Say Marie hates Nelson and wants to paint Nelson in a negative light. Marie can do this by changing the history of block A:

In [2]:
block_A.history = 'Nelson hates cat'

This is unfair to Nelson, who is a big fan of cats. So, we need to add a way in which only Nelson can write the history of his own preferences. The way to do this is by using a private key and a public key.

# 3.3 Signing data in blockchain

In blockchain, we use two keys to sign data, to authenticate a message and protect it from being altered by unauthorized users. The two keys are as follows:

- Private key
- Public key

The secrecy of the private key is guarded and it is not made known to the public. On the other hand, you let the public key be given out in public. You tell everyone, hey, this is my public key.

Let's generate the private key. 

So, Nelson generates the private key, which is the nelsonkey.pem file. He must keep this key secret. It is generated as follows:

[https://www.bitaddress.org/bitaddress.org-v3.3.0-SHA256-dec17c07685e1870960903d8f58090475b25af946fe95a734f88408cef4aa194.html]

In [3]:
from OpenSSL import crypto

key = crypto.PKey()
key.generate_key(crypto.TYPE_RSA, 1024)
key_contents = crypto.dump_privatekey(crypto.FILETYPE_PEM, key)

# Let's look at the content of the key
print(key_contents)

# Let's save the key
with open('nelsonkey.pem', 'wb') as private_key:
    private_key.write(key_contents)

b'-----BEGIN PRIVATE KEY-----\nMIICdwIBADANBgkqhkiG9w0BAQEFAASCAmEwggJdAgEAAoGBANcs/+zvVPnGL5Wp\ncRmYRsrd9KCa5u+iE4TuF6UGbiqrdU26U8nH+7Qw8Bbl31uzNrd2GyUbVpJ+TUB9\nVY3rDDlFKAafooAlLAU6j34ecLK9XbgppaT5I7oWPZlwOKSuJGst/2eg0t4LxXZy\nMQzs2lPt9GiFY7UdIz3sDDrZbMd3AgMBAAECgYBF3V+Pb7oYi2fh4S1O0SmTLvQB\nvanj/qz1yVS2CVt10Uv6G9ieuKcaP5WE6chW/t5yPQ/dQX4ffJDaspc3yCykR0vm\nqvgjPWV6pzbEkBH+pf2VLX+/hO5OMTMi0vS7P6J7d+gUn+ig8QWHPO2teQMq7yjF\n13aLoVSaYadWjivCmQJBAOz4+H4d1N+zX77+mJHh0pz9VfgGk64nTzBG06o85jYe\nXF+Ympn9d1mCBk4JZL2jDSqR0RVugpFlI1ejmOi+6GUCQQDoc/2/K4skpJ4o4umI\nfCvV4UcfuHzyayVnOq9rWZGKwjQrMwbSb+aDmls25lBQ/YGEVggTz3ryoDYN+1HV\ngpyrAkB4RjlhUWTNEW6kMvZe/J34HkC2/SeRedQgbaJawkQZ7bj2sC6vv80Gug7X\nJr9h3NL+6zmSxh8xCoMwpREj+FldAkEAnvfsGZfawYV+0DFWSku44rz8mfneIR8b\nZjxhCQYpHX7wlHaB/NogB/JA3Qum86xarbj0LzAeQd1E+kzcSyrWnwJBAKt1bYr8\nQtUP/0iLylgJhecxtQJqqHBHorXLiwjR2vtZsRjAR+K8SbUhX9AOskCvMlmfxLdb\nuWHXa+UNrvkiJKw=\n-----END PRIVATE KEY-----\n'


Another way to generate the private key and quickly the public key is:

In [4]:
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.serialization import load_pem_private_key

def gen_private_key(key_size):
    """
     key_size: currently 1024 and below are considered breakable 
     while 2048 or 4096 are reasonable default key sizes for new keys
     public_exponent: The public_exponent indicates what one mathematical property of the key generation will be. 
                         Unless you have a specific reason to do otherwise, you should always use 65537.
     "backend: An optional backend which implements RSABackend."
    """
    private_key = rsa.generate_private_key(
        public_exponent=65537, 
        key_size=key_size, 
        backend=default_backend()
    )
    return private_key

def gen_public_key(private_key):
    
    # Create public key
    public_key = private_key.public_key()
    
    return public_key

def save_key(pk, filename, type_key="private"):
    if type_key=="private":
        key_content = pk.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        )
        
    else:
        key_content = pk.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
        
    with open(filename, 'wb') as out:
        out.write(key_content)
        
def load_key(filename):
    with open(filename, 'rb') as pem_in:
        pemlines = pem_in.read()
    private_key = load_pem_private_key(pemlines, None, default_backend())
    return private_key

In [5]:
# Generate and save the private key
private_key = gen_private_key(2048)
filename = './data/nelsonkey.pem'
save_key(private_key, filename)

In [6]:
# Generate and save the public key
public_key = gen_public_key(private_key)
print("Public key generated from private key:", public_key)
filename = './data/nelsonkey.pub'
save_key(public_key, filename, type_key="public")

Public key generated from private key: <cryptography.hazmat.backends.openssl.rsa._RSAPublicKey object at 0x0000018877180390>


Let's take a look at a case in which Marie tries to falsify the facts with a script named falsify_message.py. Marie tries to put "Nelson hates cat" in the history database as follows:

In [7]:
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization

message = b'Nelson hates cat'
signature = b'Fake Signature'

with open("./data/nelsonkey.pub", "rb") as key_file:
    public_key = serialization.load_pem_public_key(
        key_file.read(),
        backend=default_backend())

public_key.verify(
 signature,
 message,
 padding.PSS(mgf=padding.MGF1(hashes.SHA256()),
                salt_length=padding.PSS.MAX_LENGTH),
    hashes.SHA256())

InvalidSignature: 

Here's how the verify method works. Nelson calculates the hash from the message, then encrypts it with his private key. The result is the signature. For example, if Sky wants to verify the signature, he has the message and the signature. He calculates the hash of the message. Then, he decrypts the signature using the public key. The result is compared to the hash of the message. If it is the same, then everything is well. If not, either the message has been altered or the private key used to sign the message is different.

Every message has a different signature, and it's impossible for Marie to guess the signature in order to falsify the message. So, with the private key and the public key, we can verify whether or not the message is indeed from someone authorized, even if we communicate on an unsecured channel.

So with the private key, Nelson could create a signature that is unique to the message it tries to sign:
<img src="./figures/nelson_key.png" width="70%"  align="center">

Everyone in the world who has Nelson's public key can verify that Nelson did indeed write Message A. Nelson can prove he did write Message A by showing Signature A. Everyone can take those two inputs and verify the truth:

<img src="./figures/nelson_key2.png" width="70%"  align="center">

## 3.4 Merkle hash trees

A Merkle tree is a binary tree where all the leaf nodes represent hashes of the data blocks. Each parent node has the hashed value of the hashes of its children. Hashing continues until the root node of the tree. Merkle trees are used to summarize bulk sets of data and create a fingerprint for each set.

> *A tree is a data structure in computer science that consists of a root node and a subtree of parent and children nodes and is represented by positioning a root node at the top. A binary tree is a tree where each parent has at most two nodes. Merkle trees are used in Bitcoin, Ethereum, and other blockchain applications to summarize all the transactions included in each block. SHA-256 is used as a hash function in bitcoin's Merkle tree, as can be seen in the following diagram*:

<img src="./figures/merkle_tree.png" width="70%"  align="center">

A Merkle tree is constructed from the bottom up from the leaf nodes. Leaf nodes will consist only of hashed values of data blocks A, B, C, and D represented by HA, HB, HC, and HD respectively. Each parent node will construct its hash by concatenating the hash values of the child nodes and hashing them again:

        HAB = Hash (HA + HB)
        
This process is continued until the root node hash value HABCD is calculated.
Since each Merkle tree node (other than leaf nodes) calculates its hash based on its child nodes, it has to maintain a balanced tree, that is, each node (other than leaf nodes) should have two child nodes. This could be achieved by duplicating existing single child nodes.

Merkle trees not only provide a way of summarizing an entire data block, but they can also efficiently verify whether a data block exists. Verification could be achieved in just `log2(n)` complexity.

[https://github.com/bitcoin/bitcoin/tree/master/src/consensus]

## 3.5 From linked list to blockchain

To make the tutorial code short, we are not going to integrate the validation using the private key and the public key. We assume only authorized people are able to write the history in the block. Let's upgrade our linked list to the blockchain. To do this, we add a new property in the Block class, which is the parent's hash:

In [ ]:
import hashlib
import json

class Block:
    id = None
    history = None
    parent_id = None
    parent_hash = None

block_A = Block()
block_A.id = 1
block_A.history = 'Nelson likes cat'

block_B = Block()
block_B.id = 2
block_B.history = 'Marie likes dog'
block_B.parent_id = block_A.id
block_B.parent_hash = hashlib.sha256(json.dumps(block_A.__dict__).encode('utf-8')).hexdigest()

block_C = Block()
block_C.id = 3
block_C.history = 'Marie likes dog'
block_C.parent_id = block_B.id
block_C.parent_hash = hashlib.sha256(json.dumps(block_B.__dict__).encode('utf-8')).hexdigest()

# 4. Hashes

A transaction in a Blockchain network is determined by its participants as correct and should be ticked off. If most of the network participants consider the transaction to be correct (by applying a consensus mechanism), the
transaction, together with a series of other transactions, is built into a block. This block is cryptographically linked (chained) to the previous block, which is also composed of a set of previous transactions.

The way the blocks are linked with each other involves so-called `“hash”` functions and ensures that every block is
linked to the previous one. `Hashing` is a function that takes an input of any length and turns it into a fixed length output. So, to make this clearer, we can look at the following code example:

In [ ]:
import hashlib
print(hashlib.sha256(b"Hello world!").hexdigest())
print(hashlib.sha256(b"Hello").hexdigest())

c0535e4be2b79ffd93291305436bf889314e4a3faec05ecffcbb7df31ad9e51a
185f8db32271fe25f561a6fc938b2e264306ec304eda518007d1764826381969


As you can see, the length of the input can be 1, 5, or even 20 characters, but the output will always be the length of 64 hexadecimal numeric characters. The output looks scrambled and it appears that there is no apparent link between the input and the output. However, if you give the same input, it will give the same output every time:

In [ ]:
print(hashlib.sha256(b"Hello world!").hexdigest())

c0535e4be2b79ffd93291305436bf889314e4a3faec05ecffcbb7df31ad9e51a


The output of the hashing function can be used as a digital signature. Imagine you have a string with a length of 10 million (say you are writing a novel), and to make sure this novel is not tampered with, you tell all your potential readers that they have to count the 10 million characters in order to ensure that the novel isn't be corrupted. Nobody would do that. But with hashing, you can publish the output validation with only 64 characters (through Twitter, for example) and your potential readers can hash the novel that they buy/download and compare them to make sure that their novel is legit.

So, we add the parent's hash in the block class. This way, we keep the digital signature of the parent's block in our block. This means that if we are ever naughty and change the content of any block, the parent's hash in any child's block will be invalid, and you would get caught red-handed.

## 4.1 How do hashes work?

A `hash` is something like the unique digital fingerprint of any imaginable set of data, regardless of its size. Technically, a hash is comparable to a cross sum. A `hash formula`, however, is much more complex than just adding
numbers for a cross sum – a *hash can be thought of a cross sum mixer that factors, adds and multiplies every single digit (including letters) of a dataset and calculates a certain result from it*: the `hash`.

*A hash is a function that converts one value to another. Hashing data is a common practice in computer science and is used for several different purposes. Examples include cryptography, compression, checksum generation, and data indexing.* (https://techterms.com/definition/hash)

As with cross sums, a `hash` can be much shorter than the original hashed text (data). It is also impossible to conclude from the `hash` back to the initial dataset – and that is a feature! For example, the hash of the phrase “nothing is decided until everything is decided” always has the hash:
`9f62f85d500c8d4682c2aa9f8a00d89658be956b3a680dfd370eb1c9bb94e445`.

A change of just one minor part of the dataset causes the so-called “avalanche effect” in the “hash mixer” and
leads to a completely different `hash`. For example, the slightly altered phrase “nothing is decided until everything’s decided” has the `hash`:
`3f9801bc00d0a466b42c006dbbbf312ce38d1cf515a999bb09f9b556feeb56244`.

In [ ]:
import hashlib

print("Available hash algorithms are:", hashlib.algorithms_guaranteed) 

Available hash algorithms are: {'sha384', 'sha512', 'sha3_256', 'shake_256', 'blake2s', 'blake2b', 'sha3_224', 'sha3_512', 'sha3_384', 'sha224', 'sha256', 'shake_128', 'sha1', 'md5'}


In [ ]:
# Using the hash algorithm used by Bitcoin
value = "COMP0163 Digital Finance"
hashed_value = hashlib.sha256(value.encode()) 
  
# printing the equivalent hexadecimal value. 
print("Hexadecimal equivalent of '{}' in SHA256 is:".format(value), hashed_value.hexdigest()) 

Hexadecimal equivalent of 'COMP0163 Digital Finance' in SHA256 is: d7574ec8acb5054dd642ebbe3e400c880753418e4c716002d1c20c95e8c998b3


<img src="./figures/blockchain_hashes.png" width="70%"  align="center">

For example, from the [different hash functions algorithms](https://en.bitcoinwiki.org/wiki/List_of_hash_functions), *Bitcoin uses the SHA-256 hash algorithm to generate verifiably "random" numbers in a way that requires a predictable amount of CPU effort. Generating a SHA-256 hash with a value less than the current target solves a block and wins you some coins.*

Transaction data is permanently recorded in files called blocks. They can be thought of as the individual pages of a city recorder's recordbook (where changes to title to real estate are recorded) or a stock transaction ledger. Blocks are organized into a linear sequence over time (also known as the blockchain). New transactions are constantly being processes by miners into new blocks which are added to the end of the chain and can never be changed or removed once accepted by the network (although some software will remove orphaned blocks). 

> *We can consider the hash as a unique identifier of a fixed size that represents each block; no two blocks will have the same identifier. Since all the blocks are linked together by this hash value, each block will have the identifier of the previous block. The previous block is referred to as the parent block, and each block can have only one parent.*

> *The block hash is not a hash value of the entire block, but rather the hash value of the block header, which consists only of metadata.*

Maintaining an open distributed blockchain with a consensus for validating the block was the main thing that made Bitcoin a practically implementable decentralized currency. Bitcoin made use of a consensus algorithm called the **proof-of-work (PoW)** algorithm to prove that a node has actually worked to create a new block of the blockchain.

In terms of cryptocurrency, the nodes that perform **PoW** are called `mining nodes`. As the name suggests, mining is the act of *forging new blocks to be appended to the blockchain*. The amount of work that it takes to mine ensures that the blockchain is **immutable** and that tampering with any past transactions is nearly impossible.

## 4.2 Proof-of-Work

So, we have three participants in this case: Nelson, Marie, and Sky. But there is another type of participant too: the one who writes into the blockchain is called—in blockchain parlance— `the miner`. `Miners` are required to create a block that satisfies certain criteria set by the network. So, in order to put the transaction into the blockchain, `the miner` is required to do some work first.

Previously, we had three blocks (block_A, block_B, and block_C), but now we have a candidate block (block_D), which we want to add into the blockchain as follows:

In [ ]:
block_D = Block()
block_D.id = 4
block_D.history = 'Sky loves turtle'
block_D.parent_id = block_C.id

But instead of adding `block_D` to the blockchain just like that, we first require the `miner` to do some puzzle work. We serialize that `block` and ask the `miner` to apply an extra string, which, when appended to the serialization string of that block, will show the `hash` output with at least five zeros in the front, if it is hashed.

First things first, we serialize the block:

In [ ]:
import json
block_serialized = json.dumps(block_D.__dict__).encode('utf-8')
print(block_serialized)
b'{"history": "Sky loves turtle", "parent_id": 3, "id": 4}'

b'{"id": 4, "history": "Sky loves turtle", "parent_id": 3}'


b'{"history": "Sky loves turtle", "parent_id": 3, "id": 4}'

If the serialized block is hashed, what does it mean if we want the hash output to have at least five zeros at the front? It means that we want the output to look like this:

`00000aa21def23ee175073c6b3c89b96cfe618b6083dae98d2a92c919c1329be`

So, the puzzle is something like this:

`string serialization + answer = hash output with (at least) 5 leading zeros`

The miner needs to guess the correct answer. If this puzzle is converted to Python code, it would be something like this:

In [ ]:
answer = ?
input = b'{"history": "Sky loves turtle", "parent_id": 3, "id": 4}' + answer
output = hashlib.sha256(input).hexdigest()

# output needs to be 00000??????????????????????????????????????????????????????????? 

SyntaxError: invalid syntax (742090172.py, line 1)

So, how could the miner solve a problem like this? We can use brute force:

In [ ]:
import hashlib

payload = b'{"history": "Sky loves turtle", "parent_id": 3, "id": 4}'

for i in range(10000000):
    
  nonce = str(i).encode('utf-8')
  result = hashlib.sha256(payload + nonce).hexdigest()
    
   # We are validating that the result has 5 zeros in the first part 
  if result[0:5] == '00000':
    print(i)
    print(result)
    break

184798
00000ae01f4cd7806e2a1fccd72fb18679cb07ede3a2a7ef028a0ecfd4aec153


This means that the answer is `184798`, or the hash output of `{"history": "Sky loves turtle", "parent_id": 3, "id": 4}`, `184798` is the one that has five leading zeros. In that simple script, we iterate from `0 to 9999999` and append that into the input. This is a naive method, but it works. Of course, you could also append with characters other than numbers, such as a, b, or c.

Now, try to increase the number of leading zeros to six, or even ten. In this case, can you find the hash output? If there is no output, you could increase the range limit from `10000000` to an even higher number, such as `1000000000000`. Once you get an appreciation of the hard work that goes into this, try to comprehend this: *Bitcoin required around 19 leading zeros in the hash output as of 2020*. The number of leading zeros is not static and changes according to the situation (but you don't need to worry about this).

## 4.3 Consensus

As we can see, the hashing function makes history tampering hard, but not too hard. Even if we have a blockchain that consists of 1000 blocks, it would be trivial to alter the content of the first block and change the 999 parent hashes on the other blocks with recent computers. So, to ensure that bad people cannot alter the history (or at least make it very hard), we distribute this append-only database to everyone who wants to keep it (let's call them miners). Say there are ten miners. In this case, you cannot just alter the blockchain in your copy because the other nine miners who would scold, saying something like *hey, our records say history A but your record says B*. In this case, the majority wins.

However, consensus is not just a case of choosing which blockchain has been chosen by the majority. The problem starts when we want to add a new block to the blockchain. Where do we start? How do we do it? The answer is that we broadcast. When we broadcast the candidate block that contains a new transaction, it will not reach every miner at the same time. You may reach the miner that stands beside you, but it will require time for your message to reach the miner that stands far away from you.

Most miners want to get that reward we spoke of in the most efficient manner possible. To do this, they would get a new candidate block, work hard to find the answer in proof of work, and then add it to the longest chain as quickly as possible. But, why do they want to add it to the longest chain and not another chain? This is because it secures their reward.

Say we have two versions of the blockchain. One has three blocks, while the other has eight blocks. The most sensible way to add a new block is to add it to the blockchain that has eight blocks. If someone adds it to the blockchain that has three blocks, it is more likely to get discarded. Consequently, the reward would be taken away from the miner. The longest chain attracts the most miners anyway, and you want to be in the blockchain version that is kept by more people.

Some miners could persist in adding the block to the blockchain with three blocks, while other miners could also persist in adding the block to the blockchain with eight blocks. We call this a hard fork. Most of the time, miners will stick to the blockchain that has the longest chain.

To change the history, an attacker will need to outgun at least more than 50% of the miners, which is impossible. The older the block, the more secure the history in that block is. Say one person needs 5 minutes to do the puzzle work. In this case, to replace the last five blocks in the blockchain, an attacker needs more than 25 minutes (because the attacker needs at least *six blocks to convince miners to replace the last five blocks in their blockchain*). 

But in those 25 minutes, other miners would keep adding new blocks to the **most popular blockchain**. So when 25 minutes have passed, the most popular blockchain would have gained an additional five blocks! Maybe the miners take a nap for an hour and don't add any more blocks. In this case, the attacker could accumulate six blocks to tamper with the most popular blockchain. However, the incentive embedded in the blockchain keeps the miners awake 24/7 as they want to get the reward as much as possible. Consequently, it's a losing battle for the attacker.

# 5. Cryptography

The most popular use of blockchain is to create a cryptocurrency. As the word crypto is in cryptocurrency, you would expect that you need to master cryptography in order to become a blockchain programmer. That is not true. You only need to know two things about cryptography:

- Private key and public key (asymmetric cryptography)
- Hashing

These two have been explained previously. You don't need to know how to design a hashing algorithm or private key and public key algorithm. You only need to get an intuitive understanding of how they work and the implications of these technologies.

> *The implication of private keys and public keys is that it enables decentralized accounts. In a normal application, you have a username and password. These two fields enable someone to access their account. But having a private key and public key enables someone to have an account in a decentralized manner.*

For hashing, it is a one-way function, meaning that given an input, you can get the output easily. But given an output, you couldn't get the input. A simple version of a one-way function would be this:

<img src="./figures/cryptography_function.png" width="70%"  align="center">

This is an addition process. If I tell you one of the outputs of this function is 999, and I ask you what the inputs are, you couldn't guess the answer. It could be anything from 1 and 998 to 500 and 499. A hashing function is something like that. The algorithm is clear as sky (you can read the algorithm of any hashing function on the internet), but it's hard to reverse the algorithm.

So, all you need to know about hashing is this: E.g. given an input you get an SHA-256 output (in hexadecimal) like: `c96c6d5be8d08a12e7b5cdc1b207fa6b2430974c86803d8891675e76fd992c20`. 

If you don't know the input, you couldn't get the input based on this output alone.

## Further readings

- [Bitcoin white paper](https://bitcoin.org/bitcoin.pdf)

- Ethereum papers:
    - [yellow paper](https://ethereum.github.io/yellowpaper/paper.pdf)
    - [White paper](https://github.com/ethereum/wiki/wiki/White-Paper)

- [IPFS documentation](https://docs.ipfs.io/reference/api/http/)

    - https://github.com/ipfs/py-ipfs-api
    - https://docs.ipfs.io/
 

## Credits

- N. Vadgama (code)

- Hands-On Blockchain for Python Developers, 2019, Arjuna Sky Kok

- Foundations of Blockchain, 2019, Koshik Raj

- [Bitcoin white paper](https://bitcoin.org/bitcoin.pdf)

- [IPFS documentation](https://docs.ipfs.io/reference/api/http/)

- Hands-On Bitcoin Programming with Python, 2018, Harish Kumar Garg

- Blockchain Potentials and Limitations for Selected Climate Policy Instruments, 2019, Sven Braden

<h1>
    <br style ="color:#EA7600"> UCL </br> 
    </p>
</h1>

<div>

  <h4>
      <p style="font-size:18pt">COMP0163 Blockchain Technologies</p>
      <p style="font-size:18pt">Lab Practical 1</p>
      <p style="font-size:18pt">Part 2</p>
  </h4>

</div>

# Table of Contents
* [1. Review](#1.-Review)
* [2. Build a public blockchain](#2.-Build-a-public-blockchain)
    * [2.1 Implement a proof of work algorithm](#2.1-Implement-a-proof-of-work-algorithm)
    * [2.2 Add blocks to the blockchain](#2.2-Add-blocks-to-the-blockchain)
    * [2.3 Mining](#2.3-Mining)
* [3. Build a Blockchain application](#3.-Build-a-Blockchain-application)
    * [3.1 Our Blockchain as an API](#3.1-Our-Blockchain-as-an-API)
    * [3.2 Consensus and decentralization](#3.2-Consensus-and-decentralization) 
    * [3.3 Build the application](#3.3-Build-the-application)
    * [3.4 Execute the blockchain application](#3.4-Execute-the-blockchain-application)
    * [3.5 Signing the messages](#3.5-Signing-the-messages)    


# 2. Build a public blockchain

After reviewing some of the main concepts of what a blockchain entails, let's build our own. We are going to build a public blockchain like the type used by Bitcoin and Ethereum.

We are going to use the class that we built in the last part to link the blocks:

In [ ]:
from hashlib import sha256
import json
import time

class Block:
    
    def __init__(self, index, transactions, timestamp, previous_hash):
        """
        Constructor for the `Block` class.
        :param index:         Unique ID of the block.
        :param transactions:  List of transactions.
        :param timestamp:     Time of generation of the block.
        :param previous_hash: Hash of the previous block in the chain which this block is part of.                                        
        """
        self.index = index
        self.transactions = transactions
        self.timestamp = timestamp
        self.previous_hash = previous_hash # Adding the previous hash field

    def compute_hash(self):
        """
        Returns the hash of the block instance by first converting it
        into JSON string.
        """
        block_string = json.dumps(self.__dict__, sort_keys=True) # The string equivalent also considers the previous_hash field now
        return sha256(block_string.encode()).hexdigest()

We are going to add another class that defines the genesis block as well as helping in retrieving the latest block:

In [ ]:
class Blockchain:

    def __init__(self):
        """
        Constructor for the `Blockchain` class.
        """
        self.chain = []
        self.create_genesis_block()

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        """
        A quick python way to retrieve the most recent block in the chain. Note that
        the chain will always consist of at least one block (i.e., genesis block)
        """
        return self.chain[-1]

Now, if the content of any of the previous blocks changes:

- The hash of that previous block would change.
- This will lead to a mismatch with the `previous_hash` field in the next block.
- Since the input data to compute the hash of any block also consists of the `previous_hash` field, the hash of the next block will also change.

Ultimately, the entire chain following the replaced block is invalidated, and the only way to fix it is to recompute the entire chain.

## 2.1 Implement a proof of work algorithm

A Proof of Work algorithm (PoW) is how new Blocks are created or mined on the blockchain. The goal of PoW is to discover a number which solves a problem. The number must be difficult to find but easy to verify—computationally speaking—by anyone on the network. This is the core idea behind Proof of Work.

To add a new block to the blockchain, we are going to require the `miner` to do some puzzle work. Here’s how we do this: Instead of accepting any hash for the block, we add some constraint to it. Let’s add a constraint that our hash should start with `“n leading zeroes”` where n can be any positive integer.

We know that unless we change the data of the block, the hash is not going to change, and of course we don’t want to change existing data. So what do we do? Simple! We’ll add some dummy data that we can change. Let’s introduce a new field in our block called nonce. A nonce is a number that we can keep on changing until we get a hash that satisfies our constraint. The nonce satisfying the constraint serves as proof that some computation has been performed. This technique is a simplified version of the [`Hashcash`](https://en.wikipedia.org/wiki/Hashcash) algorithm used in Bitcoin.

In [ ]:
class Blockchain:

    def __init__(self):
        """
        Constructor for the `Blockchain` class.
        """
        self.chain = []
        self.create_genesis_block()

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        """
        A quick pythonic way to retrieve the most recent block in the chain. Note that
        the chain will always consist of at least one block (i.e., genesis block)
        """
        return self.chain[-1]
    
    def proof_of_work(self, block):
        """
        Function that tries different values of the nonce to get a hash
        that satisfies our difficulty criteria.
        """
        block.nonce = 0

        computed_hash = block.compute_hash()
        while not computed_hash.startswith('0' * Blockchain.difficulty):
            block.nonce += 1
            computed_hash = block.compute_hash()

        return computed_hash

Notice that there is no specific logic to figuring out the nonce quickly; it’s just brute force. The only definite improvement that you can make is to use hardware chips that are specially designed to compute the hash function in a smaller number of CPU instructions.

## 2.2 Add blocks to the blockchain

To add a block to the chain, we’ll first have to verify that:

- The data has not been tampered with (the proof of work provided is correct).
- The order of transactions is preserved (the `previous_hash` field of the block to be added points to the hash of the latest block in our chain).

In [ ]:
class Blockchain:

    def __init__(self):
        """
        Constructor for the `Blockchain` class.
        """
        self.chain = []
        self.create_genesis_block()

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        """
        A quick pythonic way to retrieve the most recent block in the chain. Note that
        the chain will always consist of at least one block (i.e., genesis block)
        """
        return self.chain[-1]
    
    def proof_of_work(self, block):
        """
        Function that tries different values of the nonce to get a hash
        that satisfies our difficulty criteria.
        """
        block.nonce = 0

        computed_hash = block.compute_hash()
        while not computed_hash.startswith('0' * Blockchain.difficulty):
            block.nonce += 1
            computed_hash = block.compute_hash()

        return computed_hash
        
        
    def add_block(self, block, proof):
        """
        A function that adds the block to the chain after verification.
        Verification includes:
        * Checking if the proof is valid.
        * The previous_hash referred in the block and the hash of a latest block
          in the chain match.
        """
        previous_hash = self.last_block.hash

        if previous_hash != block.previous_hash:
            return False

        if not Blockchain.is_valid_proof(block, proof):
            return False

        block.hash = proof
        self.chain.append(block)
        return True

    def is_valid_proof(self, block, block_hash):
        """
        Check if block_hash is valid hash of block and satisfies
        the difficulty criteria.
        """
        return (block_hash.startswith('0' * Blockchain.difficulty) and
                block_hash == block.compute_hash())


## 2.3 Mining

The transactions will be initially stored as a pool of unconfirmed transactions. The process of putting the unconfirmed transactions in a block and computing proof of work is known as the `mining` of blocks. Once the nonce satisfying our constraints is figured out, we can say that a block has been mined and it can be put into the blockchain.

In most of the cryptocurrencies (including Bitcoin), miners may be awarded some cryptocurrency as a reward for spending their computing power to compute a proof of work. Here’s what our mining function looks like:

In [ ]:
class Blockchain:

    def __init__(self):
        """
        Constructor for the `Blockchain` class.
        """
        self.chain = []
        self.create_genesis_block()
        
        # difficulty of PoW algorithm
        self.difficulty = 2

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        """
        A quick pythonic way to retrieve the most recent block in the chain. Note that
        the chain will always consist of at least one block (i.e., genesis block)
        """
        return self.chain[-1]
    
    def proof_of_work(self, block):
        """
        Function that tries different values of the nonce to get a hash
        that satisfies our difficulty criteria, which is to start
        with to leading zeros.
        """
        block.nonce = 0

        computed_hash = block.compute_hash()
        
        
        while not computed_hash.startswith('0' * self.difficulty):
            block.nonce += 1
            computed_hash = block.compute_hash()

        return computed_hash    
        
    """
    Add blocks to the blockchain
    """
    
    def add_block(self, block, proof):
        """
        A function that adds the block to the chain after verification.
        Verification includes:
        * Checking if the proof is valid.
        * The previous_hash referred in the block and the hash of a latest block
          in the chain match.
        """
        previous_hash = self.last_block.hash

        if previous_hash != block.previous_hash:
            return False

        if not Blockchain.is_valid_proof(block, proof):
            return False

        block.hash = proof
        self.chain.append(block)
        return True

    def is_valid_proof(self, block, block_hash):
        """
        Check if block_hash is valid hash of block and satisfies
        the difficulty criteria.
        """
        return (block_hash.startswith('0' * Blockchain.difficulty) and
                block_hash == block.compute_hash())

    """
    Mining functions
    """
    def add_new_transaction(self, transaction):
        self.unconfirmed_transactions.append(transaction)

    def mine(self):
        """
        This function serves as an interface to add the pending
        transactions to the blockchain by adding them to the block
        and figuring out proof of work.
        """
        if not self.unconfirmed_transactions:
            return False

        last_block = self.last_block

        new_block = Block(index=last_block.index + 1,
                          transactions=self.unconfirmed_transactions,
                          timestamp=time.time(),
                          previous_hash=last_block.hash)

        proof = self.proof_of_work(new_block)
        self.add_block(new_block, proof)
        self.unconfirmed_transactions = []
        return new_block.index

# 3. Build a Blockchain application

In this part, we are going to a build a blockchain application that runs on top of the public blockchain that we just define before. The first step is to create an endpoint interface from which we can interact with our blockchain.

## 3.1 Our Blockchain as an API

We’ll be using a popular Python microframework called `Flask` to create a `REST API` that interacts with and invokes various operations in our blockchain node. It’s a micro-framework and it makes it easy to map endpoints to Python functions. This allows us talk to our blockchain over the web using HTTP requests.

We’ll create three methods:

- `/new_transaction`  to create a new transaction to a block

- `/mine` to tell our server to mine a new block.

- `/chain` to return the full Blockchain

In [ ]:
from flask import Flask, request
import requests

# Initialize flask application
app =  Flask(__name__)

# Initialize a blockchain object.
blockchain = Blockchain()

We need an endpoint for our application to submit a new transaction. This will be used by our application to add new data (posts) to the blockchain:

In [ ]:
# Flask's way of declaring end-points
@app.route('/new_transaction', methods=['POST'])
def new_transaction():
    tx_data = request.get_json()
    required_fields = ["author", "content"]

    for field in required_fields:
        if not tx_data.get(field):
            return "Invalid transaction data", 404

    tx_data["timestamp"] = time.time()

    blockchain.add_new_transaction(tx_data)

    return "Success", 201

Here’s an endpoint to return the node’s copy of the chain. Our application will be using this endpoint to query all of the data to display:

In [ ]:
@app.route('/chain', methods=['GET'])
def get_chain():
    chain_data = []
    for block in blockchain.chain:
        chain_data.append(block.__dict__)
    return json.dumps({"length": len(chain_data),
                       "chain": chain_data})

### The Mining Endpoint

Our mining endpoint is where the magic happens. It has to do three things:

- Calculate the Proof of Work
- Reward the miner (us) by adding a transaction granting us 1 coin
- Forge the new Block by adding it to the chain

Here’s an endpoint to request the node to mine the unconfirmed transactions (if any). We’ll be using it to initiate a command to mine from our application itself:

In [ ]:
@app.route('/mine', methods=['GET'])
def mine_unconfirmed_transactions():
    result = blockchain.mine()
    if not result:
        return "No transactions to mine"
    return "Block #{} is mined.".format(result)

@app.route('/pending_tx')
def get_pending_tx():
    return json.dumps(blockchain.unconfirmed_transactions)

These REST endpoints can be used to play around with our blockchain by creating some transactions and then mining them.

## 3.2 Consensus and decentralization

Up to this point, the blockchain that we’ve implemented is meant to run on a single computer. Even though we’re linking block with hashes and applying the proof of work constraint, we still can’t trust a single entity (in our case, a single machine). We need the data to be distributed, we need multiple nodes maintaining the blockchain. So, to transition from a single node to a peer-to-peer network, let’s first create a mechanism to let a new node become aware of other peers in the network:

In [ ]:
# Contains the host addresses of other participating members of the network
peers = set()

# Endpoint to add new peers to the network
@app.route('/register_node', methods=['POST'])
def register_new_peers():
    # The host address to the peer node 
    node_address = request.get_json()["node_address"]
    if not node_address:
        return "Invalid data", 400

    # Add the node to the peer list
    peers.add(node_address)

    # Return the blockchain to the newly registered node so that it can sync
    return get_chain()


@app.route('/register_with', methods=['POST'])
def register_with_existing_node():
    """
    Internally calls the `register_node` endpoint to
    register current node with the remote node specified in the
    request, and sync the blockchain as well with the remote node.
    """
    node_address = request.get_json()["node_address"]
    if not node_address:
        return "Invalid data", 400

    data = {"node_address": request.host_url}
    headers = {'Content-Type': "application/json"}

    # Make a request to register with remote node and obtain information
    response = requests.post(node_address + "/register_node",
                             data=json.dumps(data), headers=headers)

    if response.status_code == 200:
        global blockchain
        global peers
        # update chain and the peers
        chain_dump = response.json()['chain']
        blockchain = create_chain_from_dump(chain_dump)
        peers.update(response.json()['peers'])
        return "Registration successful", 200
    else:
        # if something goes wrong, pass it on to the API response
        return response.content, response.status_code


def create_chain_from_dump(chain_dump):
    blockchain = Blockchain()
    for idx, block_data in enumerate(chain_dump):
        block = Block(block_data["index"],
                      block_data["transactions"],
                      block_data["timestamp"],
                      block_data["previous_hash"])
        proof = block_data['hash']
        if idx > 0:
            added = blockchain.add_block(block, proof)
            if not added:
                raise Exception("The chain dump is tampered!!")
        else:  # the block is a genesis block, no verification needed
            blockchain.chain.append(block)
    return blockchain

A new node participating in the network can invoke the register_with_existing_node method (via the /register_with endpoint) to register with existing nodes in the network. This will help with the following:

- Asking the remote node to add a new peer to its list of known peers.
- Initializing the blockchain of the new node with that of the remote node.
- Resyncing the blockchain with the network if the node goes off-grid.


However, there’s a problem with multiple nodes. Due to intentional manipulation or unintentional reasons (like network latency), the copy of chains of a few nodes can differ. In that case, the nodes need to agree upon some version of the chain to maintain the integrity of the entire system. In other words, we need to achieve **`consensus`**.

> *A simple consensus algorithm could be to `agree upon the longest valid chain when the chains of different participating nodes in the network appear to diverge`. The rationale behind this approach is that the longest chain is a good estimate of the most amount of work done (remember proof of work is difficult to compute)*:

In [ ]:
class Blockchain:

    def __init__(self):
        """
        Constructor for the `Blockchain` class.
        """
        self.chain = []
        self.create_genesis_block()
        
        # difficulty of PoW algorithm
        self.difficulty = 2

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        """
        A quick pythonic way to retrieve the most recent block in the chain. Note that
        the chain will always consist of at least one block (i.e., genesis block)
        """
        return self.chain[-1]
    
    def proof_of_work(self, block):
        """
        Function that tries different values of the nonce to get a hash
        that satisfies our difficulty criteria, which is to start
        with to leading zeros.
        """
        block.nonce = 0

        computed_hash = block.compute_hash()
        
        
        while not computed_hash.startswith('0' * self.difficulty):
            block.nonce += 1
            computed_hash = block.compute_hash()

        return computed_hash    
        
    """
    Add blocks to the blockchain
    """
    
    def add_block(self, block, proof):
        """
        A function that adds the block to the chain after verification.
        Verification includes:
        * Checking if the proof is valid.
        * The previous_hash referred in the block and the hash of a latest block
          in the chain match.
        """
        previous_hash = self.last_block.hash

        if previous_hash != block.previous_hash:
            return False

        if not Blockchain.is_valid_proof(block, proof):
            return False

        block.hash = proof
        self.chain.append(block)
        return True

    def is_valid_proof(self, block, block_hash):
        """
        Check if block_hash is valid hash of block and satisfies
        the difficulty criteria.
        """
        return (block_hash.startswith('0' * Blockchain.difficulty) and
                block_hash == block.compute_hash())

    """
    Mining functions
    """
    def add_new_transaction(self, transaction):
        self.unconfirmed_transactions.append(transaction)

    def mine(self):
        """
        This function serves as an interface to add the pending
        transactions to the blockchain by adding them to the block
        and figuring out proof of work.
        """
        if not self.unconfirmed_transactions:
            return False

        last_block = self.last_block

        new_block = Block(index=last_block.index + 1,
                          transactions=self.unconfirmed_transactions,
                          timestamp=time.time(),
                          previous_hash=last_block.hash)

        proof = self.proof_of_work(new_block)
        self.add_block(new_block, proof)
        self.unconfirmed_transactions = []
        return new_block.index
    
    
    """
    Validate the chain being used to append blocks
    """
    
    def check_chain_validity(cls, chain):
        """
        A helper method to check if the entire blockchain is valid.            
        """
        result = True
        previous_hash = "0"

        # Iterate through every block
        for block in chain:
            block_hash = block.hash
            # remove the hash field to recompute the hash again
            # using `compute_hash` method.
            delattr(block, "hash")

            if not cls.is_valid_proof(block, block.hash) or \
                    previous_hash != block.previous_hash:
                result = False
                break

            block.hash, previous_hash = block_hash, block_hash

        return result

In [ ]:
def consensus():
    """
    Our simple consensus algorithm. If a longer valid chain is
    found, our chain is replaced with it.
    """
    global blockchain

    longest_chain = None
    current_len = len(blockchain.chain)

    for node in peers:
        response = requests.get('{}/chain'.format(node))
        length = response.json()['length']
        chain = response.json()['chain']
        if length > current_len and blockchain.check_chain_validity(chain):
              # Longer valid chain found!
            current_len = length
            longest_chain = chain

    if longest_chain:
        blockchain = longest_chain
        return True

    return False

Next, we need to develop a way for any node to announce to the network that it has mined a block so that everyone can update their blockchain and move on to mine other transactions. Other nodes can simply verify the proof of work and add the mined block to their respective chains (remember that verification is easy once the nonce is known):

In [ ]:
# endpoint to add a block mined by someone else to
# the node's chain. The node first verifies the block
# and then adds it to the chain.
@app.route('/add_block', methods=['POST'])
def verify_and_add_block():
    block_data = request.get_json()
    block = Block(block_data["index"],
                  block_data["transactions"],
                  block_data["timestamp"],
                  block_data["previous_hash"])

    proof = block_data['hash']
    added = blockchain.add_block(block, proof)

    if not added:
        return "The block was discarded by the node", 400

    return "Block added to the chain", 201


def announce_new_block(block):
    """
    A function to announce to the network once a block has been mined.
    Other blocks can simply verify the proof of work and add it to their
    respective chains.
    """
    for peer in peers:
        url = "{}add_block".format(peer)
        requests.post(url, data=json.dumps(block.__dict__, sort_keys=True))

The `announce_new_block` method should be called after every block is mined by the node so that peers can add it to their chains.

In [ ]:
@app.route('/mine', methods=['GET'])
def mine_unconfirmed_transactions():
    result = blockchain.mine()
    if not result:
        return "No transactions to mine"
    else:
        # Making sure we have the longest chain before announcing to the network
        chain_length = len(blockchain.chain)
        consensus()
        if chain_length == len(blockchain.chain):
            # announce the recently mined block to the network
            announce_new_block(blockchain.last_block)
        return "Block #{} is mined.".format(blockchain.last_block.index)

The final script would look like:

In [8]:
%%writefile ./node_server.py

from hashlib import sha256
import json
import time

from flask import Flask, request
import requests


class Block:
    def __init__(self, index, transactions, timestamp, previous_hash):
        self.index = index
        self.transactions = transactions
        self.timestamp = timestamp
        self.previous_hash = previous_hash
        self.nonce = 0

    def compute_hash(self):
        """
        A function that return the hash of the block contents.
        """
        block_string = json.dumps(self.__dict__, sort_keys=True)
        return sha256(block_string.encode()).hexdigest()


class Blockchain:
    # difficulty of our PoW algorithm
    difficulty = 2

    def __init__(self):
        self.unconfirmed_transactions = []
        self.chain = []
        self.create_genesis_block()

    def create_genesis_block(self):
        """
        A function to generate genesis block and appends it to
        the chain. The block has index 0, previous_hash as 0, and
        a valid hash.
        """
        genesis_block = Block(0, [], time.time(), "0")
        genesis_block.hash = genesis_block.compute_hash()
        self.chain.append(genesis_block)

    @property
    def last_block(self):
        return self.chain[-1]

    def add_block(self, block, proof):
        """
        A function that adds the block to the chain after verification.
        Verification includes:
        * Checking if the proof is valid.
        * The previous_hash referred in the block and the hash of latest block
          in the chain match.
        """
        previous_hash = self.last_block.hash

        if previous_hash != block.previous_hash:
            return False

        if not self.is_valid_proof(block, proof):
            return False

        block.hash = proof
        self.chain.append(block)
        return True

    def is_valid_proof(self, block, block_hash):
        """
        Check if block_hash is valid hash of block and satisfies
        the difficulty criteria.
        """
        return (block_hash.startswith('0' * Blockchain.difficulty) and
                block_hash == block.compute_hash())

    def proof_of_work(self, block):
        """
        Function that tries different values of nonce to get a hash
        that satisfies our difficulty criteria.
        """
        block.nonce = 0

        computed_hash = block.compute_hash()
        while not computed_hash.startswith('0' * Blockchain.difficulty):
            block.nonce += 1
            computed_hash = block.compute_hash()

        return computed_hash

    def add_new_transaction(self, transaction):
        self.unconfirmed_transactions.append(transaction)

    def mine(self):
        """
        This function serves as an interface to add the pending
        transactions to the blockchain by adding them to the block
        and figuring out Proof Of Work.
        """
        if not self.unconfirmed_transactions:
            return False

        last_block = self.last_block

        new_block = Block(index=last_block.index + 1,
                          transactions=self.unconfirmed_transactions,
                          timestamp=time.time(),
                          previous_hash=last_block.hash)

        proof = self.proof_of_work(new_block)
        self.add_block(new_block, proof)

        self.unconfirmed_transactions = []
        return new_block.index


app = Flask(__name__)

# the node's copy of blockchain
blockchain = Blockchain()

# the address to other participating members of the network
peers = set()


# endpoint to submit a new transaction. This will be used by
# our application to add new data (posts) to the blockchain
@app.route('/new_transaction', methods=['POST'])
def new_transaction():
    tx_data = request.get_json()
    required_fields = ["author", "content"]

    for field in required_fields:
        if not tx_data.get(field):
            return "Invlaid transaction data", 404

    tx_data["timestamp"] = time.time()

    blockchain.add_new_transaction(tx_data)

    return "Success", 201


# endpoint to return the node's copy of the chain.
# Our application will be using this endpoint to query
# all the posts to display.
@app.route('/chain', methods=['GET'])
def get_chain():
    chain_data = []
    for block in blockchain.chain:
        chain_data.append(block.__dict__)
    return json.dumps({"length": len(chain_data),
                       "chain": chain_data})


# endpoint to request the node to mine the unconfirmed
# transactions (if any). We'll be using it to initiate
# a command to mine from our application itself.
@app.route('/mine', methods=['GET'])
def mine_unconfirmed_transactions():
    result = blockchain.mine()
    if not result:
        return "No transactions to mine"
    return "Block #{} is mined.".format(result)


# endpoint to add new peers to the network.
@app.route('/add_nodes', methods=['POST'])
def register_new_peers():
    nodes = request.get_json()
    if not nodes:
        return "Invalid data", 400
    for node in nodes:
        peers.add(node)

    return "Success", 201


# endpoint to add a block mined by someone else to
# the node's chain. The block is first verified by the node
# and then added to the chain.
@app.route('/add_block', methods=['POST'])
def validate_and_add_block():
    block_data = request.get_json()
    block = Block(block_data["index"],
                  block_data["transactions"],
                  block_data["timestamp",
                  block_data["previous_hash"]])

    proof = block_data['hash']
    added = blockchain.add_block(block, proof)

    if not added:
        return "The block was discarded by the node", 400

    return "Block added to the chain", 201


# endpoint to query unconfirmed transactions
@app.route('/pending_tx')
def get_pending_tx():
    return json.dumps(blockchain.unconfirmed_transactions)


def consensus():
    """
    Our simple consnsus algorithm. If a longer valid chain is
    found, our chain is replaced with it.
    """
    global blockchain

    longest_chain = None
    current_len = len(blockchain)

    for node in peers:
        response = requests.get('http://{}/chain'.format(node))
        length = response.json()['length']
        chain = response.json()['chain']
        if length > current_len and blockchain.check_chain_validity(chain):
            current_len = length
            longest_chain = chain

    if longest_chain:
        blockchain = longest_chain
        return True

    return False


def announce_new_block(block):
    """
    A function to announce to the network once a block has been mined.
    Other blocks can simply verify the proof of work and add it to their
    respective chains.
    """
    for peer in peers:
        url = "http://{}/add_block".format(peer)
        requests.post(url, data=json.dumps(block.__dict__, sort_keys=True))

if __name__ == '__main__':
    app.run(host='0.0.0.0', debug=True, port=8000)

Overwriting ./node_server.py


## 3.3 Build the application

Our application needs to connect to a node in the blockchain network to fetch the data and also to submit new data. There can also be multiple nodes, as well.

In [9]:
import datetime
import json

import requests
from flask import render_template, redirect, request

from app import app

# Node in the blockchain network that our application will communicate with
# to fetch and add data.
CONNECTED_NODE_ADDRESS = "http://127.0.0.1:8000"

posts = []

The `fetch_posts` function gets the data from the node’s `/chain` endpoint, parses the data, and stores it locally.

In [10]:
def fetch_posts():
    """
    Function to fetch the chain from a blockchain node, parse the
    data, and store it locally.
    """
    get_chain_address = "{}/chain".format(CONNECTED_NODE_ADDRESS)
    response = requests.get(get_chain_address)
    if response.status_code == 200:
        content = []
        chain = json.loads(response.content)
        for block in chain["chain"]:
            for tx in block["transactions"]:
                tx["index"] = block["index"]
                tx["hash"] = block["previous_hash"]
                content.append(tx)

        global posts
        posts = sorted(content,
                       key=lambda k: k['timestamp'],
                       reverse=True)


The application has an HTML form to take user input and then makes a POST request to a connected node to add the transaction into the unconfirmed transactions pool. The transaction is then mined by the network, and then finally fetched once we refresh our web page:

In [11]:
@app.route('/submit', methods=['POST'])
def submit_textarea():
    """
    Endpoint to create a new transaction via our application
    """
    post_content = request.form["content"]
    author = request.form["author"]

    post_object = {
        'author': author,
        'content': post_content,
    }

    # Submit a transaction
    new_tx_address = "{}/new_transaction".format(CONNECTED_NODE_ADDRESS)

    requests.post(new_tx_address,
                  json=post_object,
                  headers={'Content-type': 'application/json'})

    # Return to the homepage
    return redirect('/')


AssertionError: View function mapping is overwriting an existing endpoint function: submit_textarea

The final script for the app would look like:

In [12]:
%%writefile ./app/views.py

import datetime
import json

import requests
from flask import render_template, redirect, request

from app import app

# The node with which our application interacts, there can be multiple
# such nodes as well.
CONNECTED_NODE_ADDRESS = "http://127.0.0.1:8000"

posts = []


def fetch_posts():
    """
    Function to fetch the chain from a blockchain node, parse the
    data and store it locally.
    """
    get_chain_address = "{}/chain".format(CONNECTED_NODE_ADDRESS)
    response = requests.get(get_chain_address)
    if response.status_code == 200:
        content = []
        chain = json.loads(response.content)
        for block in chain["chain"]:
            for tx in block["transactions"]:
                tx["index"] = block["index"]
                tx["hash"] = block["previous_hash"]
                content.append(tx)

        global posts
        posts = sorted(content, key=lambda k: k['timestamp'],
                       reverse=True)


@app.route('/')
def index():
    fetch_posts()
    return render_template('index.html',
                           title='YourNet: Decentralized '
                                 'content sharing',
                           posts=posts,
                           node_address=CONNECTED_NODE_ADDRESS,
                           readable_time=timestamp_to_string)


@app.route('/submit', methods=['POST'])
def submit_textarea():
    """
    Endpoint to create a new transaction via our application.
    """
    post_content = request.form["content"]
    author = request.form["author"]

    post_object = {
        'author': author,
        'content': post_content,
    }

    # Submit a transaction
    new_tx_address = "{}/new_transaction".format(CONNECTED_NODE_ADDRESS)

    requests.post(new_tx_address,
                  json=post_object,
                  headers={'Content-type': 'application/json'})

    return redirect('/')


def timestamp_to_string(epoch_time):
    return datetime.datetime.fromtimestamp(epoch_time).strftime('%H:%M')

Overwriting ./app/views.py


Let's define what app is and the script to call that we have defined before:

In [13]:
%%writefile ./app/__init__.py

from flask import Flask

app = Flask(__name__)

from app import views

Overwriting ./app/__init__.py


## 3.4 Execute the blockchain application

Let's run the application that is going to allow us to create transactions (post messages via the web inteface), and once we mine the transactions, all the nodes in the network will update the chain.

First, let's start blockchain node server. If you are in an Unix based operating system (Mac OS, Linux, etc.), you can execute the application by running:
        
        export FLASK_APP=./node_server.py
        flask run --port 8000
        
and in Windows it would be like:

        set FLASK_APP=./node_server.py
        flask run --port 8000

Let's create a small script to run our app:

In [14]:
%%writefile run_app.py

from app import app

app.run(debug=True)

Overwriting run_app.py


One instance of our blockchain node is now up and running at port 8000, Run the application on a different terminal session:

            python run_app.py

### If the above gives warning that "No module named 'flask' is found", please follow below
* After opening a new terminal, make sure to activate the environment again
> conda activate blockchain_env
* After activating, update your path by running the following command line (Mac OS, Linux, etc.)
> export PATH=$CONDA_PREFIX/bin:$PATH
* In Windows, the commands would be like
> set PATH=%CONDA_PREFIX%\Scripts;%PATH%
* Then, please run the application
> python run_app.py





The application should be up and running at http://localhost:5000 from where we can interact directly using an HTML web interface to simulate the operations that happen in a blockchain.

To play around by spinning off multiple custom nodes, use the `register_with/` endpoint to register a new node. Remember that you already have a node running in port 8000:

        flask run --port 8000
        
You can spin more nodes in different ports:

        flask run --port 8001
        flask run --port 8002

You can use the following `cURL requests` to register the node at port `8001` with the already running node at port `8000`:
        
        curl -X POST \
      http://127.0.0.1:8001/register_with \
      -H 'Content-Type: application/json' \
      -d '{"node_address": "http://127.0.0.1:8000"}'
  
which you can do directly in Python using the requests library:

In [17]:
import requests

headers = {
    'Content-Type': 'application/json',
}

# Data of original node already running
data = '{"node_address": "http://127.0.0.1:8000"}'

# Post request to add new node to node already running
response = requests.post('http://127.0.0.1:8001/register_with', headers=headers, data=data)

ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=8001): Max retries exceeded with url: /register_with (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x0000018878FBE810>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))

: 

The request above will make the node at port `8000` aware of the node at port `8001` and make the newer node sync the chain with the node at port `8000`. Then, they are able to actively participate in the mining process post registration.

Apart from the web interface, we can use the API to get the chain of the nodes by doing a get request to the `/chain` endpoint:

In [16]:
import requests

response = requests.get('http://localhost:8000/chain')
response.json()

{'length': 1,
 'chain': [{'index': 0,
   'transactions': [],
   'timestamp': 1697807547.5281427,
   'previous_hash': '0',
   'nonce': 0,
   'hash': 'f405fba9f3e910ca833119c4977167cb291aa8b7de5ea6c7e0d434c0dc7def40'}]}

## 3.5 Signing the messages

The previous blockchain that we built and the application running on top of it, lacks a fundamental element, which is the use of private and public keys. In blockchain, we use two keys to sign data, to authenticate a message and protect it from being altered by unauthorized users. The two keys are as follows:

- Private key
- Public key

The secrecy of the private key is guarded and it is not made known to the public. On the other hand, you let the public key be given out in public. You tell everyone, hey, this is my public key.

> The implication of private keys and public keys is that it enables decentralized accounts. In a normal application, you have a username and password. These two fields enable someone to access their account. But having a private key and public key enables someone to have an account in a decentralized manner.

*For our purposes of understanding and building an a Blockchain application, the previous example suffice enough without the signing of the messages, but have in mind that it is a must requirement for the enablement of decentralization in a blockchain.*

## Further readings

- [Bitcoin white paper](https://bitcoin.org/bitcoin.pdf)

- Ethereum papers:
    - [yellow paper](https://ethereum.github.io/yellowpaper/paper.pdf)
    - [White paper](https://github.com/ethereum/wiki/wiki/White-Paper)

## Credits

- [Develop a blockchain application from scratch in Python, 2018, Satwik Kansal](https://github.com/satwikkansal/python_blockchain_app/tree/ibm_blockchain_post)

- Learn Blockchains by Building One, 2017, Daniel van Flymen

- Hands-On Blockchain for Python Developers, 2019, Arjuna Sky Kok

- [Vyper's documentation](http://vyper.readthedocs.io)

- [Truffle's documentation](https://www.trufflesuite.com/docs/truffle/overview)

- Foundations of Blockchain, 2019, Koshik Raj

- [Bitcoin white paper](https://bitcoin.org/bitcoin.pdf)

- Hands-On Bitcoin Programming with Python, 2018, Harish Kumar Garg

- Blockchain Potentials and Limitations for Selected Climate Policy Instruments, 2019, Sven Braden